docs/src/adr/014adr-precise-records.md

```diff
@@ -0,0 +1,668 @@
+shape. As a result, when we use variants, we would have to add extra
+fields to all of them. This solution is not very different from the current
+implementation of record type checking, though it would allow us to quickly
+detect spelling errors.
+
+**Blocker.** This does not look like a real solution, as it would immediately
+render the existing examples invalid. Moreover, they would be no obvious way to
+repair these examples.
+
+### 4.2. Support plain records and variants, but no row typing
+
+In this scenario, the type checker would issue an error, if a record expression
+accesses a field that is outside of its type:
+
+```tla
+FieldAccess ==
+  LET m == [ a |-> 2, b |-> "B" ] IN
+  /\ m.a > 1        \* type OK
+  /\ m.b = "B"      \* type OK
+  /\ m.c = { 1, 2 } \* should flag a type error
+```
+
+Moreover, the following example would require a type annotation:
+
+```tla
+RowAccess(m) ==
+    m.a > 0         \* should flag a type error
+```
+
+In the above example, the type checker would not be able to infer the type of
+`m` and would require an explicit type annotation:
+
+
+```tla
+\* @type: [ a: Int, b: Str ];
+RowAccess(m) ==
+    m.a > 0         \* should not flag a type error
+```
+
+**Blocker.** It is not clear to me, what type we would assign to the record
+access operator. Although the type of `m` in `FieldAccess` is obvious to a
+human reader, as it simply requires us to do the top-to-the-bottom type
+propagation, Snowcat constructs a set of equality constraints that are solved
+by unification. This approach requires that we capture the record access
+operator `.` as an equality over type variables and types.
+
+### 4.3. Support plain records and variants, including row typing
+
+This solution follows the approach outlined by [Leijen05][], except that we
+require field names to be unique, not scoped. We present our approach in the
+next section.
+
+## 5. Solution
+
+In the following, we present row types as type terms. We discuss the
+user-facing syntax of the type system later in the text.
+
+### 5.1. Plain records
+
+By using [Row types][], we should be able to infer the type of `m` in
+`RowAccess`:
+
+```
+Rec(RowCons("a", Int, z))
+```
+
+In this example, `RowCons("a", Int, z)` is a type term indicating that the type
+has the field called `a`, and its type is `Int`. On top of that, this row
+extends a parametric type `z`, which either contains a non-empty sequence of
+rows, or is an empty sequence, that is, `RowNil`. Importantly, `RowCons` is
+wrapped with the term `Rec`, so no additional fields can be added to the type.
+
+The example `FieldAccess` contains a record constructor
+`[ a |-> 2, b |-> "B" ]`. We can write a general type inference rule for it:
+
+```
+e_1: t_1, ..., e_n: t_n
+-------------------------------- [rec]
+[ f_1 |-> e_1, ..., f_n |-> e_n]:
+  Rec(
+    RowCons(f_1, t_1,
+      RowCons(f_2, t_2,
+        ...
+          RowNil)
+      )...)
```

Comment:
Ooooh now I see it, ok, thanks! I'm not sure about the notation between `)...)` and `...))`. The first one seems to make it more clear that the dots are about parenthesis, although I didn't get it at first :stuck_out_tongue:

Reply:
For parenthesis opening and closing we have the following.
```
Rec(RowCons(f_1, t_1,RowCons(f_2, t_2,...,RowNil
    )      )                 ...        )      )
```
So the second to last closing is matching the `...`. I read it in terms of position, if you read `)...)` as an arbitrary number of `)` then I think it works as well, we just have to agree on the semantics :-).

Looking at this again, I think it would be good to have `RowCons(f_n, t_n,RowNil)` instead of simply `RowNil`.

Reply:
Yes! That would be much better. So we'd be left with some like this:
```
e_1: t_1, ..., e_n: t_n
-------------------------------- [rec]
[ f_1 |-> e_1, ..., f_n |-> e_n]:
  Rec(
    RowCons(f_1, t_1,
      RowCons(f_2, t_2,
        ...
          RowCons(f_n, t_n, RowNil)
      ...)
    )
  )
```

Reply:
fixed that